app/jobs/embargo_expiry_job.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+class EmbargoExpiryJob < Hyrax::ApplicationJob
```

Comment:
how comfortable do we feel with jobs that enqueue jobs?

could this be squashed into just one job that loops over the `records_with_expired_embargos` and lifts them directly? what's the advantage of the nested queuing?

Reply:
I suggested this approach.  I was thinking it was easier to reason about and write tests for.  Also the single object nested job would be available to be reused in other contexts.  I know that Hyrax has some jobs that enqueue other jobs in the ingest process that are linear steps, but I think this is different.  One job gathers the work to be done and the nested jobs do the work in parallel in an atomic fashion.  There might be more overhead to this but if anything fails than it might be easier to find the cause and not have it affect the rest of the work.  I'm not wed to the approach but I thought it made sense.

Can you speak more to your hesitations about nested queuing?  I'm really curious.

Reply:
> Can you speak more to your hesitations about nested queuing? I'm really curious.

i have the vague recollection that it's considered bad form form `ActiveJob` to do nested enqueues, but i can't find any actual source for that.

my concrete concern here is that it seems much harder to reason about retries. suppose `EmbargoExpiryJob` fails half-way through, after enqueuing `ExpireEmbargoJob` ~100 times, and goes back into the queue for retry. will it end up enqueuing those same ~100 jobs again (that seems like the likely outcome to me, at least for sidekiq). will the duplicate jobs fail and go into the retry queue themselves? will the reason for that be legible to someone reading worker logs?

Reply:
i do like that there's a single-object embargo/lease lift job, in this. i think it's worth keeping that utility even if the bulk expiry jobs don't reuse it.

Reply:
>my concrete concern here is that it seems much harder to reason about retries. suppose EmbargoExpiryJob fails half-way through, after enqueuing ExpireEmbargoJob ~100 times, and goes back into the queue for retry. will it end up enqueuing those same ~100 jobs again (that seems like the likely outcome to me, at least for sidekiq). will the duplicate jobs fail and go into the retry queue themselves? will the reason for that be legible to someone reading worker logs?

That's a really good point.  Is that worse than hitting an error with a specific object and having the rest behind it in line getting stuck?  What about having the atomic jobs check that they are still under embargo/lease before calling the actor?

Reply:
> That's a really good point. Is that worse than hitting an error with a specific object and having the rest behind it in line getting stuck? 

there may be complications, but i think with a single job you have the advantage that a failure still results in some legitimately partially-complete job:

  1. the bulk clearing job starts;
  1. the embargo service queries solr for a list of `n` active expired embargoes;
  1. `m` (zero or more) objects process, lifting their embargo;
  1. **failure**; re-enqueue; back-off wait
  1. re-enqueued bulk job starts;
  1.  the embargo service queries solr for a list of `n - m` active expired embargoes;
  1. ...

at least on an off-hand analysis, a job like this can be safely re-enqueued indefinitely. if it's consistently failing on the same object, it should be possible to debug that, take manual action to fix the cause, and let the job complete on a future retry. 

> What about having the atomic jobs check that they are still under embargo/lease before calling the actor?

i think this would help. if multiple jobs for the same object go into the queue, at least in this case i'd expect them all to eventually clear with successful statuses. it seems like a good idea to make sure the `Expire*Job`s behave reasonably if they are run with an object that doesn't have an embargo, or doesn't have one that's expired.

especially with the added queue wait it seems possible that the object's embargo/lease status would change between the query and the job run.

my recollection is that `EmbargoManager` is more careful about exposing and respecting the current status than the actor implementation.